---
title: "Coverage Plotting from Scratch"
author: "Peter Diakumis"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
require(tidyverse)
require(cowplot)
set.seed(1)
```

Introduction
------------
We are interested in assessing the depth of read coverage for our TruSeq and 10X
somatic COLO829 WGS samples at exonic regions of a set of cancer genes.
The `az300.bed` file contains 300 such genes of interest for AstraZeneca Oncology.
The following schematic (from
[here](https://bioconductor.org/packages/3.7/bioc/vignettes/annotatr/inst/doc/annotatr-vignette.html))
shows the basic structure of a gene.

<img src='img/gene_structure.png' alt='gene structure' width='500' />

There are a ~~few~~ lot of complexities with selecting the proper exon
coordinates for each gene, including (see [link1](https://www.biostars.org/p/16505/),
[link2](https://bioinformatics.stackexchange.com/questions/21/feature-annotation-refseq-vs-ensembl-vs-gencode-whats-the-difference)
for more):

* A gene may have several transcripts, which have different start/end coordinates
  for each exon.
* There is no standard way of representing a gene (HGNC symbol, Ensembl Gene ID,
  Entrez Gene ID?)
* There is not always a one-to-one mapping between Ensembl Gene and Entrez Gene
  IDs.

The basic idea is:

* Get the gene name for each of the az300 genes (so you only need gene name from this)
* Use the GENCODE GTF file to extract the canonical transcripts for the above gene names 
  (so you only need tx ID + gene name from this)
* Use the UCSC Exon BED file to annotate the coverage regions with transcript information 
  (can do straight with mosdepth + bedtools)

Grab only az300 canonical tx IDs from the UCSC BED file: 

```
zfgrep -f az300_tx.pattern gencode_genes_v27lift37_exons_coding_b37.bed.gz
```

Datasets
--------

* There are 6 samples per batch: 100/80/60/40/20pc + blood for TruSeq,
  TruSeq downsampled, and 10X EMA:

```{r sample_nms}
cov_fnames <- list.files("../../nogit/data/COLO829", full.names = TRUE) 
aliases <- basename(cov_fnames) %>% 
  as_tibble() %>% 
  mutate(clean = tolower(sub(".per-base.region_az300_canonical_tx.bed.gz", "", value))) %>%
  mutate(clean = sub("colo829", "", clean)) %>%
  mutate(batch = case_when(
    grepl("ema", clean) ~ "EMA",
    grepl("downsample", clean) ~ "Truseq_Down",
    grepl("truseq", clean) ~ "Truseq",
    grepl("bl", clean) ~ "EMA",
    TRUE ~ clean)) %>%
  mutate(pc = sub("_(.*)pc.*", "\\1", clean),
         pc = ifelse(grepl("b", pc), "blood", pc),
         pc = ifelse(grepl("^[-_]", pc), "100", pc)) %>% 
  unite("batch_pc", c("batch", "pc")) %>% 
  pull(batch_pc)

tibble(aliases = aliases, fnames = basename(cov_fnames)) %>% 
  arrange(aliases)
```

* A canonical transcript ID <---> az300 map has been prepared previously:

```{r az_tx_read}
az_tx <- readRDS("../../nogit/data/ref/az300_tx2gene.rds")
```


* Let's read all the coverage files into a single data frame. Below function
  simply takes the transcript column from a coverage file, separates it into
  the actual transcript ID and the exon number, uses the transcript ID to
  get the gene name from the `az_tx` object, and selects all columns
  of interest.

```{r func_def1}
read_cov <- function(fname, az_tx) {
  
  cov <- readr::read_tsv(fname, 
                         col_names = c("chromA", "startA", "endA", "depth", "chromB", "startB", "endB", "tx"),
                         col_types = "ciidciic") %>% 
    separate(tx, c('tx', 'rest'), sep = '_cds_') %>% 
    separate(rest, c('before', 'stuff'), sep = '_chr') %>% 
    separate(before, c('exon_number', 'num2'), sep = '_') %>% 
    mutate(
      depth_cat = case_when(depth < 20 ~ '< 20 Reads', TRUE ~ '>= 20 Reads'),
      depth_cat = factor(depth_cat, levels = c('< 20 Reads', '>= 20 Reads')),
      tx = case_when(grepl('_', tx) ~ gsub('_.', '', tx), TRUE ~ tx)) %>% 
    select( 
      chr = "chromA", start = "startA", end = "endA",
      tx, depth, depth_cat, exon_number,
      exon_start = "startB", exon_end = "endB")
  
  az_tx %>% 
    select(tx = transcript, gene_name) %>% 
    left_join(cov, ., by = "tx") %>% 
    select(gene_name, tx, everything())
  
}
```

```{r read_cov_files, eval=FALSE}
cov_all <- cov_fnames %>%
  set_names(aliases) %>%
  map_df(function(fn) {
    read_cov(fn, az_tx)
  }, .id = "sample")

saveRDS(cov_all, "../../nogit/data/COLO829/cov_all.rds")
```


```{r glimpse_data}
cov_all <- readRDS("../../nogit/data/COLO829/cov_all.rds")

glimpse(az_tx) # 297 tx
glimpse(cov_all) # 6,257,481 rows
```

```{r aztx2gene_func}
gene2tx <- function(gene, az_tx) {
  tx <- az_tx %>% 
    filter(gene_name == gene) %>% 
    pull(transcript) %>% 
    head(1) # CSNK2A1 has 2 tx IDs
  
  if (length(tx) == 0) stop(glue::glue("No transcript ID for {gene}"))
  
  tx
}

gene2tx("KRAS", az_tx)
gene2tx("BRCA1", az_tx)
gene2tx("FOO", az_tx)
```



```{r plot_funcs}
# set a custom color that will work even if a category is missing
# scale_colour_custom <- function(...) {
#   ggplot2:::manual_scale('colour', 
#                          values = setNames(c('darkred', 'steelblue'),
#                                            c('< 20 Reads','>= 20 Reads')), 
#                          ...)
# }
# 


plot_prep <- function(gene_name, cov) {
  transcript <- gene2tx(gene_name, az_tx)
  num_of_exons <- cov %>%
    filter(tx == transcript) %>%
    pull(exon_number) %>%
    as.numeric() %>%
    max()
  # expand to create a row for each sequence and fill in previous values
  p <- cov %>%
    filter(tx == transcript) %>% 
    group_by(sample, exon_number) %>% 
    expand(start = full_seq(c(start, end), 1)) %>% 
    left_join(., 
              cov %>%  
                filter(tx == transcript)) %>% 
    fill(gene_name:exon_end) %>% 
    ungroup() %>%
    mutate(exon_number = factor(exon_number, levels = 0:num_of_exons))
  
  p
}

plot_cov <- function(gene_name, cov) {
  plot_prep(gene_name, cov) %>% 
    ggplot(aes(x = start, xend = end, y = depth, yend = depth, colour = sample)) + 
    geom_point(size = 0.1) + 
    geom_line() +
    theme_minimal() + 
    # scale_colour_custom() +  # use my custom color set above for my three categories
    scale_y_continuous(limits = c(0, NA)) +
    theme(axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.grid.major.x = element_blank()
          # legend.position = 'none'
    ) + 
    facet_wrap(~exon_number, 
               scales = 'free_x', 
               nrow = 1, strip.position = 'bottom') +
    # facet_grid(sample ~ exon_number, scales = 'free_x') +
    ylab('Depth') + 
    xlab(gene_name)
}

```

```{r eval=FALSE}
genes <- c("KRAS", "BRCA1", "BRCA2", "TP53", "MAPK3", "CDK4", "CDK6", "TERT")
lapply(genes, function(g) {
  p <- plot_cov(g, cov_all)
  ggsave(paste0("~/Desktop/tmp/covis/", g, "_exon_covplot_all-samples.png"), p, height = 15, width = 15)
  })
```

```{r eval=FALSE}
cov <- cov_all %>% 
  filter(sample %in% c("EMA_100", "Truseq_100", "Truseq_Down_100"))
lapply(genes, function(g) {
  p <- plot_cov(g, cov)
  ggsave(paste0("~/Desktop/tmp/covis/", g, "_exon_covplot.png"), p, height = 15, width = 15)
  })
```

